File: src/main/java/aurocosh/divinefavor/common/custom_data/player/data/corrosion/ArmorCorrosionDataSerializer.java
package aurocosh.divinefavor.common.custom_data.player.data.corrosion;

import aurocosh.divinefavor.common.lib.interfaces.INbtSerializer;
import net.minecraft.nbt.NBTTagCompound;

import java.util.ArrayList;
import java.util.List;

// Handles the actual read/write of the nbt.
public class ArmorCorrosionDataSerializer implements INbtSerializer<ArmorCorrosionData> {
    private static final String TAG_CORRODED_SLOTS = "CorrodedSlots";

    @Override
    public void serialize(NBTTagCompound nbt, ArmorCorrosionData instance) {
        List<Integer> slots = instance.getCorrodedArmorSlots();
        int[] slotsArray = new int[slots.size()];
        for (int i = 0; i < slots.size(); i++)
            slotsArray[i] = slots.get(i);
        nbt.setIntArray(TAG_CORRODED_SLOTS, slotsArray);
    }

    @Override
    public void deserialize(NBTTagCompound nbt, ArmorCorrosionData instance) {
        int[] intArray = nbt.getIntArray(TAG_CORRODED_SLOTS);
        List<Integer> slots = new ArrayList<>(intArray.length);
        for (int slot : intArray)
            slots.add(slot);
        instance.setCorrodedArmorSlots(slots);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/custom_data/player/data/curse/corrosion/ArmorCorrosionDataSerializer.kt
package aurocosh.divinefavor.common.custom_data.player.data.curse.corrosion

import aurocosh.divinefavor.common.lib.interfaces.INbtSerializer
import net.minecraft.nbt.NBTTagCompound
import java.util.*

// Handles the actual read/write of the nbt.
class ArmorCorrosionDataSerializer : INbtSerializer<ArmorCorrosionData> {

    override fun serialize(nbt: NBTTagCompound, instance: ArmorCorrosionData) {
        val slots = instance.corrodedArmorSlots
        val slotsArray = IntArray(slots.size)
        for (i in slots.indices)
            slotsArray[i] = slots[i]
        nbt.setIntArray(TAG_CORRODED_SLOTS, slotsArray)
    }

    override fun deserialize(nbt: NBTTagCompound, instance: ArmorCorrosionData) {
        val intArray = nbt.getIntArray(TAG_CORRODED_SLOTS)
        val slots = ArrayList<Int>(intArray.size)
        for (slot in intArray)
            slots.add(slot)
        instance.corrodedArmorSlots = slots
    }

    companion object {
        private const val TAG_CORRODED_SLOTS = "CorrodedSlots"
    }
}
